### ISSUE TITLE
multiple dataset to train model. Sharing backbone, Separable classifier.

### ISSUE THREAD

**Issue by choieq** [PyTorchLightning/pytorch-lightning#4734]
#### What is your question?
I want to use two different dataloader to train model.
According to this two dataset, backbone network is shared and classifier is classified.
How should i? 
In training_step, i expect "batch" is list containing two dataloader but there is just one dataloader. 

#### Code

    def forward(self, x1, x2):
        x1 = self.backbone(x1)
        x2 = self.backbone(x2)

        x1 = self.avgpool1(x1)
        x1 = x1.reshape(x1.size(0), -1)
        x1 = self.fc1(x1)
        x1 = self.classifier1(x1)

        x2 = self.avgpool2(x2)
        x2 = x2.reshape(x2.size(0), -1)
        x2 = self.fc2(x2)
        x2 = self.classifier2(x2)

        return x1, x2

    def training_step(self, batch, batch_idx):
        x1, y1 = batch[0]
        x2, y2 = batch[1]
        pred1, pred2 = self(x1), self(x2)
        loss = self.loss_function(pred1, y1) + self.loss_function(pred2, y2)

        return {'loss': loss}

    def train_dataloader(self):
        img_path = '/home/choieq/pet_recog/data/Useful/Image_dataset/Training_aug2_400/'
        video_path = '/home/choieq/pet_recog/data/Useful/video/total_video/train'
        img_train_dataset = KistdogDataset(root_path=img_path, phase='train', resize=224)
        video_train_dataset = KistdogDataset(root_path=video_path, phase='train', resize=224)

        img_train_loader = DataLoader(img_train_dataset, batch_size=256, shuffle=True, num_workers=10, pin_memory=True)
        video_train_loader = DataLoader(video_train_dataset, batch_size=256, shuffle=True, num_workers=10, pin_memory=True)
        #loaders = {'image': img_train_loader, 'video': video_train_loader}
        return [img_train_loader, video_train_loader]

#### What's your environment?

 - OS: [ubuntu 18.04]
 - Packaging [conda]
 - Version [torch:1.7, pytorch-lightning:0.9.1rc2]

**Comment by SeanNaren** [comment]
Hey! See here: https://github.com/PyTorchLightning/pytorch-lightning/issues/3336#issuecomment-686896859

In progress but this is a tricky case to solve.

**Comment by rohitgr7** [comment]
you can return a dict or tuple from `KistdogDataset`'s `__getitem__` and make this work.

**Comment by EQlearning** [comment]
> you can return a dict or tuple from `KistdogDataset`'s `__getitem__` and make this work.

Thank you for reply.

Now i'm trying, but how to i solve different length of two dataset? 
In__getitem__, i got the **list index out of range** error

**Comment by EQlearning** [comment]
> Hey! See here: [#3336 (comment)](https://github.com/PyTorchLightning/pytorch-lightning/issues/3336#issuecomment-686896859)
> 
> In progress but this is a tricky case to solve.

Thank you for reply. 

But my code is not working, if I include train_dataloader in network class. 
So, I specify train_dataloader in trainer.fit().
Is a pytorch-lightning version problem?

**Comment by rohitgr7** [comment]
>  think, def len may be wrong?

looks like your video data is smaller than the image data. I'd suggest either making `len = min(len(images), len(videos))` or oversample you video data w.r.t to image data with `self.video_image_list[idx % len(video_data)]`

**Comment by EQlearning** [comment]
> > think, def len may be wrong?
> 
> looks like your video data is smaller than the image data. I'd suggest either making `len = min(len(images), len(videos))` or oversample you video data w.r.t to image data with `self.video_image_list[idx % len(video_data)]`.
> 
> Just a thought, your video data seems to contain images. Why not mix both of them and use it as a single image dataset??

Thank you, I'll try.

Right, 
Actually, video data is cropped image from video. 
But the video data has a distribution different from called image dataset.
So, I want to try train model by replacing each data loader in every epoch because quite well in image dataset. 
My goal is to perform well in video dataset.
